ocaml/xapi/xapi_host.ml

```diff
@@ -2974,10 +2974,22 @@ let get_host_updates_handler (req : Http.Request.t) s _ =
       Unixext.really_write_string s json_str |> ignore
   )
 
+let is_toolstack_restart_required ~__context self =
+  Db.Host.get_recommended_guidances ~__context ~self
+  |> List.mem `restart_toolstack
+
+let assert_master_does_not_requires_restart_toolstack ~__context =
+  if Helpers.get_master ~__context |> is_toolstack_restart_required ~__context
+  then
+    raise Api_errors.(Server_error (require_toolstack_coordinator_restart, []))
+
 let apply_updates ~__context ~self ~hash =
   (* This function runs on master host *)
   Helpers.assert_we_are_master ~__context ;
   Pool_features.assert_enabled ~__context ~f:Features.Updates ;
+  if not (Helpers.is_pool_master ~__context ~host:self) then
```

Comment:
The ref of coordinator can be get here and be compared with `self`. Then it could be used in `Db.Host.get_recommended_guidances` as well. In this way, one of `Helpers.get_master` and `Helpers.is_pool_master` is not needed.

Reply:
```
let coordinator = Helpers.get_master in
if coordinator <> self then (
  Db.Host.get_recommended_guidances ~__context ~self:coordinator
  |> List.mem `restart_toolstack
  |> (fun x -> if x then raise Api_errors.(Server_error (require_toolstack_coordinator_restart, [])))
) ;
```